control-plane/pkg/reconciler/channel/channel.go

```diff
@@ -352,6 +380,185 @@ func (r *Reconciler) finalizeKind(ctx context.Context, channel *messagingv1beta1
+		// to a single subscription, we should still set the error
+		// on the KafkaChannel status.
+		// For problems about the subscription, we should set it on the
+		// subscriber.
+		// However, core eventing-controller is setting the status
+		// of individual subscriptions, we cannot set the status of
+		// the subscription itself. We can set it on the
+		// channel.Spec.Subscribers though.
+		return statusConditionManager.FailedToGetConfig(err)
+	}
+
+	logger.Debug("Got contract config map")
+
+	// Get data plane config data.
+	ct, err := r.GetDataPlaneConfigMapData(logger, contractConfigMap)
+	if err != nil || ct == nil {
+		return statusConditionManager.FailedToGetDataFromConfigMap(err)
+	}
+
+	logger.Debug(
+		"Got contract data from config map",
+		zap.Any(base.ContractLogKey, ct),
+	)
+
+	channelIndex := coreconfig.FindResource(ct, channel.UID)
+	if channelIndex == coreconfig.NoResource {
+		return statusConditionManager.FailedToResolveConfig(errors.New("missing channel in contract"))
+	}
+
+	after := channel.DeepCopy()
+	after.Status.Subscribers = make([]v1.SubscriberStatus, 0)
+	for _, s := range channel.Spec.Subscribers {
+		err := r.reconcileSubscriber(ctx, kafkaClient, kafkaClusterAdmin, channel, s, ct, channelIndex, contractConfigMap)
+		if err != nil {
+			logger.Error("error reconciling subscription. marking subscription as not ready. ", zap.String("channel", fmt.Sprintf("%s.%s", channel.Namespace, channel.Name)), zap.Any("subscription", s))
+			after.Status.Subscribers = append(after.Status.Subscribers, v1.SubscriberStatus{
+				UID:                s.UID,
+				ObservedGeneration: s.Generation,
+				Ready:              corev1.ConditionFalse,
+				Message:            fmt.Sprintf("Subscription not ready: %v", err),
+			})
+		} else {
+			logger.Debug("marking subscription as ready. ", zap.String("channel", fmt.Sprintf("%s.%s", channel.Namespace, channel.Name)), zap.Any("subscription", s))
+			after.Status.Subscribers = append(after.Status.Subscribers, v1.SubscriberStatus{
+				UID:                s.UID,
+				ObservedGeneration: s.Generation,
+				Ready:              corev1.ConditionTrue,
+			})
+		}
+	}
+
+	jsonPatch, err := duck.CreatePatch(channel, after)
+	if err != nil {
+		return fmt.Errorf("creating JSON patch: %w", err)
+	}
+	// If there is nothing to patch, we are good, just return.
+	// Empty patch is [], hence we check for that.
+	if len(jsonPatch) == 0 {
+		logger.Debug("Patch for KafkaChannel subscriber status is empty")
+		return nil
+	}
+
+	logger.Debug("Patch for KafkaChannel subscriber status is not empty")
+
+	patch, err := jsonPatch.MarshalJSON()
+	if err != nil {
+		return fmt.Errorf("marshaling JSON patch: %w", err)
+	}
+	patched, err := r.kafkaClientSet.MessagingV1beta1().KafkaChannels(channel.Namespace).Patch(ctx, channel.Name, types.JSONPatchType, patch, metav1.PatchOptions{}, "status")
+
+	if err != nil {
+		return fmt.Errorf("failed patching: %w", err)
+	}
```

Comment:
I checked the IMC code. There, the dispatcher is responsible for marking the subscription ready/unready and it is done using a patch. This makes sense since the dispatcher is not the owner for the channel and some manual API calls are required.

I guess that code was carried to knative-sandbox/eventing-kafka and I carried it over to here.

I will now remove this code and see what happens.